File: 200.go
package main

import (
	"fmt"
	"strconv"
)

func numIslands(grid [][]byte) int {
	if len(grid) == 0 {
		return 0
	}
	m, n := len(grid), len(grid[0])
	marked := make(map[string]bool)
	num := 0
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			key := strconv.Itoa(i) + strconv.Itoa(j)
			if _, ok := marked[key]; !ok{
				if grid[i][j] == []byte("0")[0] {
					marked[key] = true
					continue
				}
				num += 1
				labers := [][]int{{i, j}}
				for len(labers) != 0 {
					r, c := labers[0][0], labers[0][1]
					labers = labers[1:]
					if 0 <= r && r < m && 0 <= c && c < n {
						_key := strconv.Itoa(r) + strconv.Itoa(c)
						if _, ok := marked[_key]; !ok{
							marked[_key] = true
							if grid[r][c] == []byte("1")[0] {
								labers = append(labers, [][]int{{r, c - 1}, {r, c + 1}, {r - 1, c}, {r + 1, c}}...)
							}
						}
					}
				}
			}
		}
	}
	return num
}

func main() {
	param := [][]byte{
		{'1', '1', '1', '1', '0'},
		{'1', '1', '0', '1', '0'},
		{'1', '1', '0', '0', '0'},
		{'0', '0', '0', '0', '0'},
	}
	a := numIslands(param)
	fmt.Print(a)
	param2 := [][]byte{
		{'1', '1', '0', '0', '0'},
		{'1', '1', '0', '0', '0'},
		{'0', '0', '1', '0', '0'},
		{'0', '0', '0', '1', '1'},
	}
	fmt.Print(numIslands(param2))
}

File: 200.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        m, n = len(grid), len(grid[0])
        marked = set()
        num = 0
        for i in range(m):
            for j in range(n):
                if (i, j) in marked:
                    continue
                if grid[i][j] == "0":
                    marked.add((i, j))
                    continue
                num += 1
                labors = [(i, j)]
                while labors:
                    r, c = labors.pop()
                    if 0 <= r < m and 0 <= c < n:
                        if (r, c) in marked:
                            continue
                        marked.add((r, c))
                        if grid[r][c] == "1":
                            labors.extend([(r, c - 1), (r, c + 1), (r - 1, c), (r + 1, c)])
        return num


if __name__ == '__main__':
    s = Solution()
    print(s.numIslands([
        ['1', '1', '1', '1', '0'],
        ['1', '1', '0', '1', '0'],
        ['1', '1', '0', '0', '0'],
        ['0', '0', '0', '0', '0']
    ]
    ))
    print(s.numIslands([
        ['1', '1', '0', '0', '0'],
        ['1', '1', '0', '0', '0'],
        ['0', '0', '1', '0', '0'],
        ['0', '0', '0', '1', '1']
    ]
    ))
